oxide/lib.go

```diff
@@ -22,71 +21,86 @@ const TokenEnvVar = "OXIDE_TOKEN"
+	// https://api.oxide.computer.
+	Address string
+
+	// Oxide API authentication token.
+	Token string
+
+	// A custom HTTP client to use for the Client instead of the default.
+	HTTPClient *http.Client
+
+	// A custom user agent string to add to every request instead of the
+	// default.
+	UserAgent string
+}
+
 // Client which conforms to the OpenAPI3 specification for this service.
 type Client struct {
-	// The endpoint of the server conforming to this interface, with scheme,
-	// https://api.oxide.computer for example.
+	// Base URL of the Oxide API including the scheme. For example,
+	// https://api.oxide.computer.
 	server string
 
-	// Client is the *http.Client for performing requests.
+	// Oxide API authentication token.
+	token string
+
+	// HTTP client to make API requests.
 	client *http.Client
 
-	// token is the API token used for authentication.
-	token string
+	// The user agent string to add to every API request.
+	userAgent string
 }
 
-// NewClient creates a new client for the Oxide API.
-// You need to pass in your API token to create the client.
-func NewClient(token, userAgent, host string) (*Client, error) {
-	if token == "" {
-		return nil, fmt.Errorf("you need to pass in an API token to create the client")
-	}
-
-	baseURL, err := parseBaseURL(host)
-	if err != nil {
-		return nil, err
+// NewClient creates a new client for the Oxide API. Pass in a non-nil *Config
+// to set the various configuration options on a Client.
+func NewClient(cfg *Config) (*Client, error) {
+	token := os.Getenv(TokenEnvVar)
+	server := os.Getenv(HostEnvVar)
+	userAgent := defaultUserAgent
+	httpClient := &http.Client{
+		Timeout: 600 * time.Second,
 	}
 
-	client := &Client{
-		server: baseURL,
-		token:  token,
-	}
+	// Layer in the user-provided configuration if provided.
+	if cfg != nil {
+		if cfg.Address != "" {
+			server = cfg.Address
+		}
 
-	// Ensure the server URL always has a trailing slash.
-	if !strings.HasSuffix(client.server, "/") {
-		client.server += "/"
-	}
+		if cfg.Token != "" {
+			token = cfg.Token
+		}
 
-	uat := userAgentTransport{
-		base:      http.DefaultTransport,
-		userAgent: userAgent,
-		client:    client,
-	}
+		if cfg.UserAgent != "" {
+			userAgent = cfg.UserAgent
+		}
 
-	client.client = &http.Client{
-		Transport: uat,
-		// We want a longer timeout since some of the files might take a bit.
-		Timeout: 600 * time.Second,
+		if cfg.HTTPClient != nil {
+			httpClient = cfg.HTTPClient
+		}
 	}
 
-	return client, nil
-}
-
-// NewClientFromEnv creates a new client for the Oxide API, using the token
-// stored in the environment variable `OXIDE_TOKEN` and the host stored in the
-// environment variable `OXIDE_HOST`.
-func NewClientFromEnv(userAgent string) (*Client, error) {
-	token := os.Getenv(TokenEnvVar)
-	if token == "" {
-		return nil, fmt.Errorf("the environment variable %s must be set with your API token", TokenEnvVar)
+	server, err := parseBaseURL(server)
```

Comment:
Before proceeding with this line, let's add some validation with friendly error messages to make sure we have an address and token.

Reply:
Updated the logic to check for a valid address and token with proper error messages.